File: web/visual_debugger/src/vocabulary.js
/**
 * Display-only vocabulary for renderer-neutral debugger scene records.
 *
 * This module maps stable semantic IDs to labels and visual keys. It does not
 * define durations, legality, ranges, acceptance, combat values, or any other
 * simulator behavior.
 */

/**
 * @typedef {"class" | "team" | "status" | "activation" | "modifier" | "lifecycle"} VisualTokenKind
 */

/**
 * @typedef {{
 *   tokenId: string,
 *   label: string,
 *   shortLabel: string,
 *   accessibleName: string,
 *   glyphKey: string,
 *   cssKey: string,
 *   fallback: string,
 * }} VisualToken
 */

/**
 * @typedef {Readonly<Record<string, Readonly<VisualToken>>>} VisualTokenRegistry
 */

/**
 * @typedef {{
 *   label?: unknown,
 *   short_label?: unknown,
 *   shortLabel?: unknown,
 *   accessible_name?: unknown,
 *   accessibleName?: unknown,
 * }} TokenPayload
 */

/**
 * @param {VisualToken} definition
 * @returns {Readonly<VisualToken>}
 */
function token(definition) {
  return Object.freeze(definition);
}

const CLASS_TOKENS = Object.freeze({
  mage: token({
    tokenId: "mage",
    label: "Mage",
    shortLabel: "Mage",
    accessibleName: "Mage class",
    glyphKey: "class-mage",
    cssKey: "mage",
    fallback: "M",
  }),
  warrior: token({
    tokenId: "warrior",
    label: "Warrior",
    shortLabel: "Warrior",
    accessibleName: "Warrior class",
    glyphKey: "class-warrior",
    cssKey: "warrior",
    fallback: "W",
  }),
  hunter: token({
    tokenId: "hunter",
    label: "Hunter",
    shortLabel: "Hunter",
    accessibleName: "Hunter class",
    glyphKey: "class-hunter",
    cssKey: "hunter",
    fallback: "H",
  }),
  rogue: token({
    tokenId: "rogue",
    label: "Rogue",
    shortLabel: "Rogue",
    accessibleName: "Rogue class",
    glyphKey: "class-rogue",
    cssKey: "rogue",
    fallback: "R",
  }),
  priest: token({
    tokenId: "priest",
    label: "Priest",
    shortLabel: "Priest",
    accessibleName: "Priest class",
    glyphKey: "class-priest",
    cssKey: "priest",
    fallback: "P",
  }),
});

const TEAM_TOKENS = Object.freeze({
  team_a: token({
    tokenId: "team_a",
    label: "Team A",
    shortLabel: "A",
    accessibleName: "Team A, solid outline",
    glyphKey: "team-a",
    cssKey: "team-a",
    fallback: "A",
  }),
  team_b: token({
    tokenId: "team_b",
    label: "Team B",
    shortLabel: "B",
    accessibleName: "Team B, solid outline",
    glyphKey: "team-b",
    cssKey: "team-b",
    fallback: "B",
  }),
});

const STATUS_TOKENS = Object.freeze({
  stun_warrior_charge: token({
    tokenId: "stun_warrior_charge",
    label: "Charge stun",
    shortLabel: "C-STN",
    accessibleName: "Warrior Charge stun",
    glyphKey: "status-charge-stun",
    cssKey: "stun-warrior-charge",
    fallback: "CS",
  }),
  stun_hunter_trap: token({
    tokenId: "stun_hunter_trap",
    label: "Trap",
    shortLabel: "TRAP",
    accessibleName: "Hunter Trap stun",
    glyphKey: "status-trap",
    cssKey: "stun-hunter-trap",
    fallback: "T",
  }),
  stun_rogue_poison: token({
    tokenId: "stun_rogue_poison",
    label: "Poison stun",
    shortLabel: "P-STN",
    accessibleName: "Rogue Poison stun",
    glyphKey: "status-poison-stun",
    cssKey: "stun-rogue-poison",
    fallback: "PS",
  }),
  slow_warrior_charge: token({
    tokenId: "slow_warrior_charge",
    label: "Charge slow",
    shortLabel: "C-SLW",
    accessibleName: "Warrior Charge slow",
    glyphKey: "status-slow",
    cssKey: "slow-warrior-charge",
    fallback: "CS",
  }),
  slow_hunter_basic: token({
    tokenId: "slow_hunter_basic",
    label: "Hunter slow",
    shortLabel: "H-SLW",
    accessibleName: "Hunter Basic slow",
    glyphKey: "status-slow",
    cssKey: "slow-hunter-basic",
    fallback: "HS",
  }),
  slow_rogue_poison: token({
    tokenId: "slow_rogue_poison",
    label: "Poison slow",
    shortLabel: "P-SLW",
    accessibleName: "Rogue Poison slow",
    glyphKey: "status-slow",
    cssKey: "slow-rogue-poison",
    fallback: "PS",
  }),
  anti_heal_rogue_poison: token({
    tokenId: "anti_heal_rogue_poison",
    label: "Anti-heal",
    shortLabel: "ANTI",
    accessibleName: "Rogue Poison anti-heal",
    glyphKey: "status-anti-heal",
    cssKey: "anti-heal-rogue-poison",
    fallback: "AH",
  }),
  priest_freedom: token({
    tokenId: "priest_freedom",
    label: "Freedom",
    shortLabel: "FREE",
    accessibleName: "Priest Blessing of Freedom",
    glyphKey: "status-freedom",
    cssKey: "priest-freedom",
    fallback: "F",
  }),
  mage_burst: token({
    tokenId: "mage_burst",
    label: "Burst",
    shortLabel: "BURST",
    accessibleName: "Mage Burst damage amplification",
    glyphKey: "status-burst",
    cssKey: "mage-burst",
    fallback: "B",
  }),
});

const ACTIVATION_TOKENS = Object.freeze({
  basic_damage: token({
    tokenId: "basic_damage",
    label: "Basic damage",
    shortLabel: "Basic",
    accessibleName: "Accepted Basic damage activation",
    glyphKey: "activation-basic-damage",
    cssKey: "basic-damage",
    fallback: "B",
  }),
  basic_heal: token({
    tokenId: "basic_heal",
    label: "Basic healing",
    shortLabel: "Basic",
    accessibleName: "Accepted Priest Basic healing activation",
    glyphKey: "activation-basic-heal",
    cssKey: "basic-heal",
    fallback: "B",
  }),
  holy_word: token({
    tokenId: "holy_word",
    label: "Holy Word",
    shortLabel: "Holy",
    accessibleName: "Accepted Priest Holy Word activation",
    glyphKey: "activation-holy-word",
    cssKey: "holy-word",
    fallback: "U",
  }),
  mage_burst: token({
    tokenId: "mage_burst",
    label: "Burst activation",
    shortLabel: "Burst",
    accessibleName: "Accepted Mage Burst activation",
    glyphKey: "activation-burst",
    cssKey: "mage-burst",
    fallback: "U",
  }),
  warrior_charge: token({
    tokenId: "warrior_charge",
    label: "Charge",
    shortLabel: "Charge",
    accessibleName: "Accepted Warrior Charge activation",
    glyphKey: "activation-charge",
    cssKey: "warrior-charge",
    fallback: "U",
  }),
  hunter_trap: token({
    tokenId: "hunter_trap",
    label: "Trap activation",
    shortLabel: "Trap",
    accessibleName: "Accepted Hunter Trap activation",
    glyphKey: "activation-trap",
    cssKey: "hunter-trap",
    fallback: "U",
  }),
  rogue_poison: token({
    tokenId: "rogue_poison",
    label: "Poison",
    shortLabel: "Poison",
    accessibleName: "Accepted Rogue Poison activation",
    glyphKey: "activation-poison",
    cssKey: "rogue-poison",
    fallback: "U",
  }),
});

const MODIFIER_TOKENS = Object.freeze({
  mage_amplification: token({
    tokenId: "mage_amplification",
    label: "Mage aura amplification",
    shortLabel: "AMP",
    accessibleName: "Effective Mage damage amplification aura modifier",
    glyphKey: "modifier-amplification",
    cssKey: "mage-amplification",
    fallback: "AMP",
  }),
  warrior_mitigation: token({
    tokenId: "warrior_mitigation",
    label: "Warrior aura mitigation",
    shortLabel: "MIT",
    accessibleName: "Effective Warrior damage mitigation aura modifier",
    glyphKey: "modifier-mitigation",
    cssKey: "warrior-mitigation",
    fallback: "MIT",
  }),
  rogue_anti_heal: token({
    tokenId: "rogue_anti_heal",
    label: "Anti-heal modifier",
    shortLabel: "ANTI",
    accessibleName: "Effective Rogue Poison anti-heal modifier",
    glyphKey: "status-anti-heal",
    cssKey: "rogue-anti-heal",
    fallback: "AH",
  }),
  priest_freedom: token({
    tokenId: "priest_freedom",
    label: "Freedom speed floor",
    shortLabel: "FREE",
    accessibleName: "Effective Priest Freedom movement-speed floor",
    glyphKey: "status-freedom",
    cssKey: "priest-freedom",
    fallback: "F",
  }),
  mage_burst: token({
    tokenId: "mage_burst",
    label: "Burst amplification",
    shortLabel: "BURST",
    accessibleName: "Effective Mage Burst damage amplification modifier",
    glyphKey: "status-burst",
    cssKey: "mage-burst",
    fallback: "B",
  }),
});

const LIFECYCLE_TOKENS = Object.freeze({
  applied: token({
    tokenId: "applied",
    label: "Applied",
    shortLabel: "Apply",
    accessibleName: "Status applied",
    glyphKey: "lifecycle-applied",
    cssKey: "applied",
    fallback: "+",
  }),
  refreshed: token({
    tokenId: "refreshed",
    label: "Refreshed",
    shortLabel: "Refresh",
    accessibleName: "Status refreshed or reapplied",
    glyphKey: "lifecycle-refreshed",
    cssKey: "refreshed",
    fallback: "R",
  }),
  decremented: token({
    tokenId: "decremented",
    label: "Aged",
    shortLabel: "Age",
    accessibleName: "Status duration decremented",
    glyphKey: "lifecycle-decremented",
    cssKey: "decremented",
    fallback: "-",
  }),
  expired: token({
    tokenId: "expired",
    label: "Expired",
    shortLabel: "Expire",
    accessibleName: "Status expired naturally",
    glyphKey: "lifecycle-expired",
    cssKey: "expired",
    fallback: "E",
  }),
  trap_broken: token({
    tokenId: "trap_broken",
    label: "Trap broken",
    shortLabel: "Break",
    accessibleName: "Hunter Trap ended by accepted damage",
    glyphKey: "lifecycle-trap-broken",
    cssKey: "trap-broken",
    fallback: "X",
  }),
  cleared_unclassified: token({
    tokenId: "cleared_unclassified",
    label: "Status ended",
    shortLabel: "End",
    accessibleName: "Status ended for an unclassified or ambiguous reason",
    glyphKey: "lifecycle-ended",
    cssKey: "cleared-unclassified",
    fallback: "?",
  }),
  trap_broken_and_reapplied: token({
    tokenId: "trap_broken_and_reapplied",
    label: "Trap broken and reapplied",
    shortLabel: "Break+",
    accessibleName: "Hunter Trap was broken and exactly reapplied",
    glyphKey: "lifecycle-trap-broken-reapplied",
    cssKey: "trap-broken-and-reapplied",
    fallback: "X+",
  }),
});

export const CLASS_TOKEN_IDS = Object.freeze([
  "mage",
  "warrior",
  "hunter",
  "rogue",
  "priest",
]);

export const TEAM_TOKEN_IDS = Object.freeze(["team_a", "team_b"]);

export const CANONICAL_STATUS_ORDER = Object.freeze([
  "stun_warrior_charge",
  "stun_hunter_trap",
  "stun_rogue_poison",
  "slow_warrior_charge",
  "slow_hunter_basic",
  "slow_rogue_poison",
  "anti_heal_rogue_poison",
  "priest_freedom",
  "mage_burst",
]);

export const ACTIVATION_TOKEN_IDS = Object.freeze([
  "basic_damage",
  "basic_heal",
  "holy_word",
  "mage_burst",
  "warrior_charge",
  "hunter_trap",
  "rogue_poison",
]);

export const MODIFIER_TOKEN_IDS = Object.freeze([
  "mage_amplification",
  "warrior_mitigation",
  "rogue_anti_heal",
  "priest_freedom",
  "mage_burst",
]);

export const LIFECYCLE_TOKEN_IDS = Object.freeze([
  "applied",
  "refreshed",
  "decremented",
  "expired",
  "trap_broken",
  "cleared_unclassified",
  "trap_broken_and_reapplied",
]);

/** @type {Readonly<Record<VisualTokenKind, VisualTokenRegistry>>} */
const TOKEN_REGISTRIES = Object.freeze({
  class: CLASS_TOKENS,
  team: TEAM_TOKENS,
  status: STATUS_TOKENS,
  activation: ACTIVATION_TOKENS,
  modifier: MODIFIER_TOKENS,
  lifecycle: LIFECYCLE_TOKENS,
});

/** @type {Readonly<Record<number, string>>} */
const CLASS_ID_TO_TOKEN = Object.freeze({
  1: "mage",
  2: "warrior",
  3: "hunter",
  4: "rogue",
  5: "priest",
});

/** @type {Readonly<Record<number, string>>} */
const TEAM_ID_TO_TOKEN = Object.freeze({
  1: "team_a",
  2: "team_b",
});

/**
 * @param {unknown} value
 * @returns {string | null}
 */
function displayString(value) {
  return typeof value === "string" && value.trim() ? value.trim() : null;
}

/**
 * @param {unknown} payload
 * @returns {TokenPayload}
 */
function tokenPayload(payload) {
  return typeof payload === "object" && payload !== null
    ? /** @type {TokenPayload} */ (payload)
    : {};
}

/**
 * @param {unknown} tokenId
 * @returns {string}
 */
function normalizedTokenId(tokenId) {
  return displayString(tokenId) ?? "unknown";
}

/**
 * @param {VisualTokenKind} kind
 * @param {string} tokenId
 * @returns {Readonly<VisualToken>}
 */
function unknownToken(kind, tokenId) {
  return token({
    tokenId,
    label: "Unknown",
    shortLabel: "?",
    accessibleName: `Unknown ${kind} visual token ${tokenId}`,
    glyphKey: "unknown",
    cssKey: "unknown",
    fallback: "?",
  });
}

/**
 * Apply Python-authored display prose without allowing payloads to select
 * glyphs or CSS hooks.
 *
 * @param {Readonly<VisualToken>} definition
 * @param {unknown} payload
 * @returns {Readonly<VisualToken>}
 */
function withPayloadProse(definition, payload) {
  const record = tokenPayload(payload);
  const label = displayString(record.label) ?? definition.label;
  const shortLabel =
    displayString(record.short_label) ??
    displayString(record.shortLabel) ??
    definition.shortLabel;
  const accessibleName =
    displayString(record.accessible_name) ??
    displayString(record.accessibleName) ??
    definition.accessibleName;

  if (
    label === definition.label &&
    shortLabel === definition.shortLabel &&
    accessibleName === definition.accessibleName
  ) {
    return definition;
  }
  return token({
    ...definition,
    label,
    shortLabel,
    accessibleName,
  });
}

/**
 * Resolve a stable semantic ID to display-only metadata.
 *
 * Python-provided labels and accessible names override registry prose. Glyph
 * and CSS keys always come from this allowlisted registry so future/unknown
 * payload IDs cannot inject selectors or markup.
 *
 * @param {VisualTokenKind} kind
 * @param {unknown} tokenId
 * @param {unknown} [payload]
 * @returns {Readonly<VisualToken>}
 */
export function resolveVisualToken(kind, tokenId, payload) {
  const normalized = normalizedTokenId(tokenId);
  const registry = TOKEN_REGISTRIES[kind];
  const definition = Object.hasOwn(registry, normalized)
    ? registry[normalized]
    : unknownToken(kind, normalized);
  return withPayloadProse(definition, payload);
}

/**
 * Resolve a simulator class identity without copying class mechanics.
 *
 * @param {unknown} classId
 * @param {unknown} [payload]
 * @returns {Readonly<VisualToken>}
 */
export function classTokenFromId(classId, payload) {
  const numericId =
    typeof classId === "number" && Number.isInteger(classId) ? classId : null;
  const tokenId = numericId === null ? undefined : CLASS_ID_TO_TOKEN[numericId];
  return resolveVisualToken(
    "class",
    tokenId ?? (numericId === null ? "unknown" : `class_id_${numericId}`),
    payload,
  );
}

/**
 * Resolve a simulator team identity without copying team mechanics.
 *
 * @param {unknown} teamId
 * @param {unknown} [payload]
 * @returns {Readonly<VisualToken>}
 */
export function teamTokenFromId(teamId, payload) {
  const numericId =
    typeof teamId === "number" && Number.isInteger(teamId) ? teamId : null;
  const tokenId = numericId === null ? undefined : TEAM_ID_TO_TOKEN[numericId];
  return resolveVisualToken(
    "team",
    tokenId ?? (numericId === null ? "unknown" : `team_id_${numericId}`),
    payload,
  );
}

File: src/marl_battlegrounds/rendering/vocabulary.py
"""Stable semantic tokens shared by debugger and replay presentation.

This module names presentation facts; it does not define simulator mechanics,
durations, legality, acceptance, or combat values.
"""

from dataclasses import dataclass
from typing import Literal

from marl_battlegrounds.core.types import (
    HUNTER_CLASS_ID,
    MAGE_CLASS_ID,
    PRIEST_CLASS_ID,
    ROGUE_CLASS_ID,
    TEAM_A_ID,
    TEAM_B_ID,
    WARRIOR_CLASS_ID,
)

type ClassTokenId = Literal["mage", "warrior", "hunter", "rogue", "priest"]
type TeamTokenId = Literal["team_a", "team_b"]
type StatusTokenId = Literal[
    "stun_warrior_charge",
    "stun_hunter_trap",
    "stun_rogue_poison",
    "slow_warrior_charge",
    "slow_hunter_basic",
    "slow_rogue_poison",
    "anti_heal_rogue_poison",
    "priest_freedom",
    "mage_burst",
]
type ActivationTokenId = Literal[
    "basic_damage",
    "basic_heal",
    "holy_word",
    "mage_burst",
    "warrior_charge",
    "hunter_trap",
    "rogue_poison",
]
type ModifierTokenId = Literal[
    "mage_amplification",
    "warrior_mitigation",
    "rogue_anti_heal",
    "priest_freedom",
    "mage_burst",
]
type StatusLifecycleKind = Literal[
    "applied",
    "refreshed",
    "decremented",
    "expired",
    "trap_broken",
    "cleared_unclassified",
    "trap_broken_and_reapplied",
]
type TokenFamily = Literal[
    "class",
    "team",
    "hard_control",
    "slow",
    "combat_modifier",
    "basic_activation",
    "ultimate_activation",
    "aura_modifier",
    "lifecycle",
    "unknown",
]


@dataclass(frozen=True, slots=True, kw_only=True)
class VisualTokenDefinition:
    """Renderer-neutral labels and fallbacks for one stable semantic token."""

    token_id: str
    label: str
    short_label: str
    accessible_name: str
    family: TokenFamily
    glyph: str
    fallback: str
    priority: int
    source_class_id: int | None

    def __post_init__(self) -> None:
        for name in (
            "token_id",
            "label",
            "short_label",
            "accessible_name",
            "glyph",
            "fallback",
        ):
            value = getattr(self, name)
            if type(value) is not str or not value.strip():
                msg = f"{name} must be a non-empty Python string."
                raise ValueError(msg)
        if type(self.priority) is not int or self.priority < 0:
            msg = f"priority must be a non-negative Python int; got {self.priority!r}."
            raise ValueError(msg)
        if self.source_class_id is not None and type(self.source_class_id) is not int:
            raise ValueError("source_class_id must be a Python int or None.")


CLASS_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="mage",
        label="Mage",
        short_label="Mage",
        accessible_name="Mage class",
        family="class",
        glyph="✦",
        fallback="M",
        priority=0,
        source_class_id=MAGE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="warrior",
        label="Warrior",
        short_label="Warrior",
        accessible_name="Warrior class",
        family="class",
        glyph="◆",
        fallback="W",
        priority=1,
        source_class_id=WARRIOR_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="hunter",
        label="Hunter",
        short_label="Hunter",
        accessible_name="Hunter class",
        family="class",
        glyph="⌖",
        fallback="H",
        priority=2,
        source_class_id=HUNTER_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="rogue",
        label="Rogue",
        short_label="Rogue",
        accessible_name="Rogue class",
        family="class",
        glyph="◈",
        fallback="R",
        priority=3,
        source_class_id=ROGUE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="priest",
        label="Priest",
        short_label="Priest",
        accessible_name="Priest class",
        family="class",
        glyph="✚",
        fallback="P",
        priority=4,
        source_class_id=PRIEST_CLASS_ID,
    ),
)

TEAM_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="team_a",
        label="Team A",
        short_label="A",
        accessible_name="Team A, solid outline",
        family="team",
        glyph="A",
        fallback="A",
        priority=0,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="team_b",
        label="Team B",
        short_label="B",
        accessible_name="Team B, solid outline with chevron marker",
        family="team",
        glyph="B",
        fallback="B",
        priority=1,
        source_class_id=None,
    ),
)

# Hard control precedes slows, which precede combat modifiers. This is the
# canonical nine-status dock order used by every renderer.
CANONICAL_STATUS_ORDER: tuple[StatusTokenId, ...] = (
    "stun_warrior_charge",
    "stun_hunter_trap",
    "stun_rogue_poison",
    "slow_warrior_charge",
    "slow_hunter_basic",
    "slow_rogue_poison",
    "anti_heal_rogue_poison",
    "priest_freedom",
    "mage_burst",
)

STATUS_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="stun_warrior_charge",
        label="Charge stun",
        short_label="C-STN",
        accessible_name="Warrior Charge stun",
        family="hard_control",
        glyph="⬢",
        fallback="CS",
        priority=0,
        source_class_id=WARRIOR_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="stun_hunter_trap",
        label="Trap",
        short_label="TRAP",
        accessible_name="Hunter Trap stun",
        family="hard_control",
        glyph="▦",
        fallback="T",
        priority=1,
        source_class_id=HUNTER_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="stun_rogue_poison",
        label="Poison stun",
        short_label="P-STN",
        accessible_name="Rogue Poison stun",
        family="hard_control",
        glyph="⬣",
        fallback="PS",
        priority=2,
        source_class_id=ROGUE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="slow_warrior_charge",
        label="Charge slow",
        short_label="C-SLW",
        accessible_name="Warrior Charge slow",
        family="slow",
        glyph="⌄",
        fallback="CS",
        priority=3,
        source_class_id=WARRIOR_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="slow_hunter_basic",
        label="Hunter slow",
        short_label="H-SLW",
        accessible_name="Hunter Basic slow",
        family="slow",
        glyph="⌄",
        fallback="HS",
        priority=4,
        source_class_id=HUNTER_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="slow_rogue_poison",
        label="Poison slow",
        short_label="P-SLW",
        accessible_name="Rogue Poison slow",
        family="slow",
        glyph="⌄",
        fallback="PS",
        priority=5,
        source_class_id=ROGUE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="anti_heal_rogue_poison",
        label="Anti-heal",
        short_label="ANTI",
        accessible_name="Rogue Poison anti-heal",
        family="combat_modifier",
        glyph="♡̸",
        fallback="AH",
        priority=6,
        source_class_id=ROGUE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="priest_freedom",
        label="Freedom",
        short_label="FREE",
        accessible_name="Priest Blessing of Freedom",
        family="combat_modifier",
        glyph="⛓̸",
        fallback="F",
        priority=7,
        source_class_id=PRIEST_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="mage_burst",
        label="Burst",
        short_label="BURST",
        accessible_name="Mage Burst damage amplification",
        family="combat_modifier",
        glyph="✷",
        fallback="B",
        priority=8,
        source_class_id=MAGE_CLASS_ID,
    ),
)

ACTIVATION_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="basic_damage",
        label="Basic damage",
        short_label="Basic",
        accessible_name="Accepted Basic damage activation",
        family="basic_activation",
        glyph="➤",
        fallback="B",
        priority=0,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="basic_heal",
        label="Basic healing",
        short_label="Basic",
        accessible_name="Accepted Priest Basic healing activation",
        family="basic_activation",
        glyph="✚",
        fallback="B",
        priority=1,
        source_class_id=PRIEST_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="holy_word",
        label="Holy Word",
        short_label="Holy",
        accessible_name="Accepted Priest Holy Word activation",
        family="ultimate_activation",
        glyph="✥",
        fallback="U",
        priority=2,
        source_class_id=PRIEST_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="mage_burst",
        label="Burst activation",
        short_label="Burst",
        accessible_name="Accepted Mage Burst activation",
        family="ultimate_activation",
        glyph="✷",
        fallback="U",
        priority=3,
        source_class_id=MAGE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="warrior_charge",
        label="Charge",
        short_label="Charge",
        accessible_name="Accepted Warrior Charge activation",
        family="ultimate_activation",
        glyph="➠",
        fallback="U",
        priority=4,
        source_class_id=WARRIOR_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="hunter_trap",
        label="Trap activation",
        short_label="Trap",
        accessible_name="Accepted Hunter Trap activation",
        family="ultimate_activation",
        glyph="▦",
        fallback="U",
        priority=5,
        source_class_id=HUNTER_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="rogue_poison",
        label="Poison",
        short_label="Poison",
        accessible_name="Accepted Rogue Poison activation",
        family="ultimate_activation",
        glyph="◆",
        fallback="U",
        priority=6,
        source_class_id=ROGUE_CLASS_ID,
    ),
)

MODIFIER_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="mage_amplification",
        label="Mage aura amplification",
        short_label="AMP",
        accessible_name="Effective Mage damage amplification aura modifier",
        family="aura_modifier",
        glyph="↑",
        fallback="AMP",
        priority=0,
        source_class_id=MAGE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="warrior_mitigation",
        label="Warrior aura mitigation",
        short_label="MIT",
        accessible_name="Effective Warrior damage mitigation aura modifier",
        family="aura_modifier",
        glyph="↓",
        fallback="MIT",
        priority=1,
        source_class_id=WARRIOR_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="rogue_anti_heal",
        label="Anti-heal modifier",
        short_label="ANTI",
        accessible_name="Effective Rogue Poison anti-heal modifier",
        family="combat_modifier",
        glyph="♡̸",
        fallback="AH",
        priority=2,
        source_class_id=ROGUE_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="priest_freedom",
        label="Freedom speed floor",
        short_label="FREE",
        accessible_name="Effective Priest Freedom movement-speed floor",
        family="combat_modifier",
        glyph="⛓̸",
        fallback="F",
        priority=3,
        source_class_id=PRIEST_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="mage_burst",
        label="Burst amplification",
        short_label="BURST",
        accessible_name="Effective Mage Burst damage amplification modifier",
        family="combat_modifier",
        glyph="✷",
        fallback="B",
        priority=4,
        source_class_id=MAGE_CLASS_ID,
    ),
)

LIFECYCLE_TOKENS: tuple[VisualTokenDefinition, ...] = (
    VisualTokenDefinition(
        token_id="applied",
        label="Applied",
        short_label="Apply",
        accessible_name="Status applied",
        family="lifecycle",
        glyph="+",
        fallback="+",
        priority=0,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="refreshed",
        label="Refreshed",
        short_label="Refresh",
        accessible_name="Status refreshed or reapplied",
        family="lifecycle",
        glyph="↻",
        fallback="R",
        priority=1,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="decremented",
        label="Aged",
        short_label="Age",
        accessible_name="Status duration decremented",
        family="lifecycle",
        glyph="-",
        fallback="-",
        priority=2,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="expired",
        label="Expired",
        short_label="Expire",
        accessible_name="Status expired naturally",
        family="lifecycle",
        glyph="⌛",
        fallback="E",
        priority=3,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="trap_broken",
        label="Trap broken",
        short_label="Break",
        accessible_name="Hunter Trap ended by accepted damage",
        family="lifecycle",
        glyph="✕",
        fallback="X",
        priority=4,
        source_class_id=HUNTER_CLASS_ID,
    ),
    VisualTokenDefinition(
        token_id="cleared_unclassified",
        label="Status ended",
        short_label="End",
        accessible_name="Status ended for an unclassified or ambiguous reason",
        family="lifecycle",
        glyph="○",
        fallback="?",
        priority=5,
        source_class_id=None,
    ),
    VisualTokenDefinition(
        token_id="trap_broken_and_reapplied",
        label="Trap broken and reapplied",
        short_label="Break+",
        accessible_name="Hunter Trap was broken and exactly reapplied",
        family="lifecycle",
        glyph="✕+",
        fallback="X+",
        priority=6,
        source_class_id=HUNTER_CLASS_ID,
    ),
)


def _lookup(
    definitions: tuple[VisualTokenDefinition, ...],
    token_id: str,
) -> VisualTokenDefinition:
    if type(token_id) is not str or not token_id.strip():
        raise ValueError("token_id must be a non-empty Python string.")
    for definition in definitions:
        if definition.token_id == token_id:
            return definition
    return VisualTokenDefinition(
        token_id=token_id,
        label="Unknown",
        short_label="?",
        accessible_name=f"Unknown visual token {token_id}",
        family="unknown",
        glyph="?",
        fallback="?",
        priority=10_000,
        source_class_id=None,
    )


def lookup_class_token(token_id: str) -> VisualTokenDefinition:
    """Return one class token definition or a safe unknown fallback."""
    return _lookup(CLASS_TOKENS, token_id)


def lookup_team_token(token_id: str) -> VisualTokenDefinition:
    """Return one team token definition or a safe unknown fallback."""
    return _lookup(TEAM_TOKENS, token_id)


def lookup_status_token(token_id: str) -> VisualTokenDefinition:
    """Return one status token definition or a safe unknown fallback."""
    return _lookup(STATUS_TOKENS, token_id)


def lookup_activation_token(token_id: str) -> VisualTokenDefinition:
    """Return one activation token definition or a safe unknown fallback."""
    return _lookup(ACTIVATION_TOKENS, token_id)


def lookup_modifier_token(token_id: str) -> VisualTokenDefinition:
    """Return one modifier token definition or a safe unknown fallback."""
    return _lookup(MODIFIER_TOKENS, token_id)


def lookup_lifecycle_token(token_id: str) -> VisualTokenDefinition:
    """Return one lifecycle token definition or a safe unknown fallback."""
    return _lookup(LIFECYCLE_TOKENS, token_id)


def class_token_from_id(class_id: int) -> VisualTokenDefinition:
    """Resolve a simulator class ID without copying class mechanics."""
    if type(class_id) is not int:
        raise ValueError("class_id must be a Python int.")
    for definition in CLASS_TOKENS:
        if definition.source_class_id == class_id:
            return definition
    return _lookup(CLASS_TOKENS, f"class_id_{class_id}")


def team_token_from_id(team_id: int) -> VisualTokenDefinition:
    """Resolve a simulator team ID without copying team mechanics."""
    if type(team_id) is not int:
        raise ValueError("team_id must be a Python int.")
    token_id = {
        TEAM_A_ID: "team_a",
        TEAM_B_ID: "team_b",
    }.get(team_id, f"team_id_{team_id}")
    return _lookup(TEAM_TOKENS, token_id)


def status_sort_key(token_id: str) -> tuple[int, str]:
    """Return the canonical status priority with deterministic unknown ordering."""
    definition = lookup_status_token(token_id)
    return definition.priority, definition.token_id


if tuple(definition.token_id for definition in STATUS_TOKENS) != (
    CANONICAL_STATUS_ORDER
):
    raise AssertionError("status-token registry order must match canonical order")
